### PR TITLE
Allow depletion without multiprocessing

### PR DESCRIPTION
**Pull request by drewejohnson**
[A user reported issues on a HPC with multiprocessing and depletion.](https://groups.google.com/forum/#!topic/openmc-users/xHKYV-EBgrY) Some environments, through networking and/or MPI restrictions, do not allow `fork()` to create necessary subprocesses during the depletion solution. This leads to the depletion hanging at
https://github.com/openmc-dev/openmc/blob/a905e5c9e22a7682c273233572904e21bc0ec739/openmc/deplete/pool.py#L53-L56

This PR allows integrators to deplete without multiprocessing, but with [`itertools.starmap`](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.pool.Pool.starmap), essentially the serial version of [`multiprocessing.Pool.starmap`](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.pool.Pool.starmap). ~All integrators support the argument
`Integrator(..., with_multiprocessing=False)` to enable or disable the use of multiprocessing.~ The default is to use multiprocessing.

~In order to facilitate this, two intermediate classes are used. One that mimics the `multiprocessing.Pool` API with respect to `starmap` and use as a context manager. The other interfaces between the `Integrator` and the `deplete` function (now a method) and uses `multiprocessing.Pool` or this new `PoolShim` class, if `multiprocessing` is to be used or not.~

The depletion regression test is also parametrized to deplete with and without multiprocessing. Using `itertools.starmap` is slower on my machine
```
$ time pytest tests/regression_tests/deplete/test.py::test_full[True]

real    0m6.550s
user    0m10.088s
sys     0m0.573s

$ time pytest tests/regression_tests/deplete/test.py::test_full[False]
real    0m8.134s
user    0m9.101s
sys     0m0.375s
```
shows the time in the regression test with and without multiprocessing, respectively. We see it is a little slower, but this will depend on computing environment and number of depletable materials. 

As for the impact on Travis for the new regression test, the individual builds do take around or slightly longer to build and test, but not dramatically. The largest increase is for `OMP=y,MPI=y,PHDF5=y,DAGMC=y` which increased just under two minutes to 46 min 37 s. Not sure if this is due to this specific test. We are still under the 50 minute Travis time limit, but I understand not wanting to get too close. 

[Develop Travis build](https://travis-ci.com/github/drewejohnson/openmc/builds/172098877)

[Feature Travis build before a minor change to `PoolShim.starmap` signature](https://travis-ci.com/github/drewejohnson/openmc/builds/172112487)

## Update
After a spirited discussion, the use of multiprocessing is handled by a boolean switch `openmc.deplete.pool.USE_MULTIPROCESSING`. It is up to the user to disable this switch as they need it, rather than pass arguments through the `Integrator`s

### CODE REVIEW

**Comment by drewejohnson** [comment]
Thanks for the comments @paulromano. I provided a very simple function to control the hidden option, but that doesn't live outside of `openmc.deplete.pool`. The implementation is definitely a lot cleaner now :rocket:

File: openmc/deplete/abc.py [hunk 1]
```diff
@@ -749,6 +761,8 @@ def __init__(self, operator, timesteps, power=None, power_density=None,
         else:
             self.solver = solver
 
+        use_multiprocessing(with_multiprocessing)
```
Comment (paulromano) [inline openmc/deplete/abc.py:764]: I think you can avoid having a function on top of the global variable:
```Python
from . import pool

...
pool._USE_MULTIPROCESSING = with_multiprocessing
```

**Comment by drewejohnson** [comment]
I think the additional option (renamed per your suggestion) is my preferred route; this feels like a more stable fix. Implementation changes can be hidden from the user, e.g. if we change the variable name / location down the road, user's don't have to update anything.

Granted, this isn't really our issue, so I don't know how much more development / interest will be put on getting a "more parallel" solution to work w/o multiprocessing

**Comment by drewejohnson** [comment]
I guess the difference in approach would be if you wanted to re-use an integrator, e.g. multiple calls to `integrate` with the same attached operator. Then the `_finalize` method would feel more at home in the `__del__` method. This might be useful if someone wanted to modify reaction rates and then run a restart simulation (?).

I'm not convinced either way is better than the other, but I think having the `_USE_MULTIPROCESSING` flag toggled symmetrically (`__init__` and `__del__` or before and after `integrate`) seems the most consistent. Putting it in the `integrate` method would allow someone to deplete with the same integrator twice in the same simulation, but with two different uses of multiprocessing. I'm not sure how common it is for the same integrator instance to be used in a simulation, and having one integration with multiprocessing and the other without seems even more unlikely...

Probably way overthinking this :sweat_smile:

**Comment by paulromano** [comment]
If you want to keep around an argument for this, I'd prefer to see it on the `integrate` method. I generally don't like to rely on `__del__` because it's pretty easy to accidentally keep a reference to an object around that prevents it from reaching `__del__`. Having the argument on `integrate` limits the scope that the global variable is changed. Also, consider the following example, which would cause issues with the constructor approach:
```Python
int1 = PredictorIntegrator(..., multiprocessing=False)
int2 = PredictorIntegrator(..., multiprocessing=True)
int1.integrate()
```

I know you're not crazy about it, but I'll make one final case for my original idea :smile: Having the user manipulate `USE_MULTIPROCESSING` directly is actually the least ambiguous option here. If someone wants to use this option, they probably _do_ want it to be persistent rather than being attached to a specific object/method call, and having them toggle a global variable seems like the cleanest way to state that intent.

**Comment by drewejohnson** [comment]
Giving it some more thought, I'm convinced of your approach. It's explicit and direct and puts the user in more control. I'll update the implementation, test, and documentation accordingly

File: docs/source/pythonapi/deplete.rst [hunk 2]
```diff
@@ -145,6 +145,17 @@ with :func:`cram.CRAM48` being the default.
    cram.CRAM48
    pool.deplete
 
+.. data:: pool.USE_MULTIPROCESSING
+
+   Boolean switch to enable or disable the use of :mod:`multiprocessing`
+   when solving the Bateman equations. The default is to use
+   :mod:`multiprocessing`, but can cause the simulation for hang in
```
Comment (paulromano) [inline docs/source/pythonapi/deplete.rst:152]:
```suggestion
   :mod:`multiprocessing`, but can cause the simulation to hang in
```

File: docs/source/pythonapi/deplete.rst [hunk 3]
```diff
@@ -145,6 +145,17 @@ with :func:`cram.CRAM48` being the default.
...
+   some computing environments, namely due to MPI and networking
+   restrictions. Disabling this option will not utilize all processing
+   units and take a little more time.
```
Comment (paulromano) [inline docs/source/pythonapi/deplete.rst:155]:
```suggestion
   restrictions. Disabling this option will result in only a single
   CPU core being used for depletion.
```

File: openmc/deplete/abc.py [hunk 4]
```diff
@@ -887,6 +887,14 @@ def integrate(self):
...
                          p, self._i_res + len(self), proc_time)
             self.operator.write_bos_data(len(self) + self._i_res)
+        self._finalize()
+
+    @staticmethod
+    def _finalize():
+        # Revert the se of multiprocessing, regardless of
+        # how the integration was performed
+        global _USE_MULTIPROCESSING
+        _USE_MULTIPROCESSING = True
```
Comment (paulromano) [inline openmc/deplete/abc.py:897]: Looks like this is leftover from before

File: openmc/deplete/abc.py [hunk 5]
```diff
@@ -1021,6 +1029,9 @@ def integrate(self):
                          p, self._i_res + len(self), proc_time)
             self.operator.write_bos_data(self._i_res + len(self))
 
+        # Tidy up: reset use of multiprocessing
+        super()._finalize()
```
Comment (paulromano) [inline openmc/deplete/abc.py:1033]: Same thing here

**Pull request merged**